src/theory/quantifiers/proof_checker.cpp

```diff
@@ -0,0 +1,104 @@
+/*********************                                                        */
+/*! \file proof_checker.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Reynolds
+ ** This file is part of the CVC4 project.
+ ** Copyright (c) 2009-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved.  See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** \brief Implementation of quantifiers proof checker
+ **/
+
+#include "theory/quantifiers/proof_checker.h"
+
+#include "expr/skolem_manager.h"
+#include "theory/builtin/proof_checker.h"
+
+using namespace CVC4::kind;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+void QuantifiersProofRuleChecker::registerTo(ProofChecker* pc)
+{
+  // add checkers
+  pc->registerChecker(PfRule::WITNESS_INTRO, this);
+  pc->registerChecker(PfRule::EXISTS_INTRO, this);
+  pc->registerChecker(PfRule::SKOLEMIZE, this);
+  pc->registerChecker(PfRule::INSTANTIATE, this);
+}
+
+Node QuantifiersProofRuleChecker::checkInternal(
+    PfRule id, const std::vector<Node>& children, const std::vector<Node>& args)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  // compute what was proven
+  if (id == PfRule::WITNESS_INTRO || id == PfRule::EXISTS_INTRO)
+  {
+    Assert(children.size() == 1);
+    Assert(args.size() == 1);
+    SkolemManager* sm = nm->getSkolemManager();
+    Node p = SkolemManager::getSkolemForm(children[0]);
+    Node t = SkolemManager::getSkolemForm(args[0]);
```

Comment:
Are these necessary or are you just being careful? I thought that these intro rules would be applied over skolem forms. Is it due to nested skolemizations?

Reply:
All external functions must call getSkolemForm.  For instance, this includes substitutions/rewrites/theory-specific proof rules.  Witness form is only used internally in the proof checker.

Reply:
This policy has been updated on proof-new recently.  The proof checker now operates on skolem form. I've updated the PR to simplify these calls.